conf/modules/optical_flow_mateksys_3901_l0x.xml

```diff
@@ -0,0 +1,50 @@
+<!DOCTYPE module SYSTEM "module.dtd">
+
+<module name="mateksys_3901_l0x">
+  <doc>
+    <description>
+      Transfers flow X and Y and range data of the Mateksys3901-l0X sensor to the autopilot
+      This by the default serial connection not by connecting the debug pins on ther sensor
+      The Data is in MSP compatible output. If a MSP stack is availabe in the AP one could use that
+      In ths ase we just handle the byres comming from the sensor in MSP compatible handler for this sensor
+    </description>
+    <configure name="MATEKSYS_3901_L0X_PORT" value="UART1" description="Set to which UART the sensor serial stream pin is connected to"/>
+    <configure name="MATEKSYS_3901_L0X_BAUD" value="115200" description="Sets the baudrate of the UART"/>
+    
+  </doc>
+
+
+  <header>
+    <file name="mateksys_3901_l0x.h" dir="modules/optical_flow"/>
+  </header>
+  <init fun="mateksys3901l0x_init()"/>
+  <event fun="mateksys3901l0x_event()"/>
+  <!-- <periodic fun="mateksys3901l0x_downlink()"/> --> <!-- to get messages TODO-->
+  <makefile>
+
+    <!-- Configure default UART port and baudrate -->
+    <configure name="MATEKSYS_3901_L0X_PORT" default="UART3" case="upper|lower"/>
+    <configure name="MATEKSYS_3901_L0X_BAUD" default="B115200"/>
+
+    <!-- Enable UART and set baudrate -->
+    <define name="USE_$(MATEKSYS_3901_L0X_PORT_UPPER)"/>
+        <!--<define name="USE_$(MATEKSYS_3901_L0X_PORT)_TX" value="FALSE"/>-->
+    <define name="$(MATEKSYS_3901_L0X_PORT_UPPER)_BAUD" value="$(MATEKSYS_3901_L0X_BAUD)"/>
+    <define name="MATEKSYS_3901_L0X_PORT" value="$(MATEKSYS_3901_L0X_PORT_LOWER)"/>
+
+    <configure name="MATEKSYS_3901_L0X_MOTION_THRES" default="120"/>
+    <configure name="MATEKSYS_3901_L0X_DISTANCE_THRES" default="200"/>
+    <configure name="USE_MATEKSYS_3901_L0X_AGL" default="0"/>
+    <configure name="MATEKSYS_3901_L0X_COMPENSATE_ROTATION" default="0"/>
```

Comment:
Do these need to be `configure`'s? If this is for the default values, I'd suggest the following instead:
1. In the C file, use the `#ifndef X; #define X; #endif` pattern to set the default values
2. In this xml file, add the defines inside the description element

Reply:
Hey thank you for the feedback I am currently working on it. Regarding this comment, I wanted to be able to change the default values directly in the airframe. If I am setting the default values in the C file, will I be able to change them later? Do I have to set a value in the xml file as well?

For example I am setting this value in the C file

```
#ifndef MATEKSYS_3901_L0X_MOTION_THRES
#define MATEKSYS_3901_L0X_MOTION_THRES 200
#endif
```

then I add this to the xml

```
  <doc>
    <description>
      Transfers flow X and Y and range data of the Mateksys3901-l0X sensor to the autopilot
      This by the default serial connection not by connecting the debug pins on ther sensor
      The Data is in MSP compatible output. If a MSP stack is availabe in the AP one could use that
      In ths ase we just handle the byres comming from the sensor in MSP compatible handler for this sensor
    </description>
    <configure name="MATEKSYS_3901_L0X_PORT" value="UART1" description="Set to which UART the sensor serial stream pin is connected to"/>
    <configure name="MATEKSYS_3901_L0X_BAUD" value="115200" description="Sets the baudrate of the UART"/>
    <define name="MATEKSYS_3901_L0X_MOTION_THRES" value="120" description="Sets the minimum motion quality to accept the flow measurement"/>
  </doc>
```

Which value will be taken at exectution? 200 or 120? Should I add something to the makefile?

Thank you

Reply:
values that differ from one drone to another should be defines.

<configure> blocks are defines in Makefiles
<define> blocks become defines in the code

So MATEKSYS_3901_L0X_PORT_LOWER is a configure: it only exists in the Makefile: the same value has to be filled at several locations.

but MATEKSYS_3901_L0X_MOTION_THRES, MATEKSYS_3901_L0X_DISTANCE_THRES etc should be defines only. So remove lines 35-38 and keep 39-42 (as an option) AND put them in the code as @tomvand suggests.